### DISCUSSION TITLE
Roadmap: As-code

### DISCUSSION THREAD

**Discussion by sdboyer** [grafana/grafana#39593] [Platform]
**This roadmap is now defunct and should not be relied upon.** It's a bad-news, great-news situation.

Bad news: it's out of date and inaccurate because changing internal priorities at Grafana Labs meant approaching things as outlined in this roadmap didn't really make sense.

Great news: it no longer made sense to approach things as outlined here because the changes to our internal priorities mean that producing clear, consumable schema for Grafana objects is now seen as an essential prerequisite to much larger, unifying long-term goals. That demands something more robust than a scattershot roadmap. And boy howdy, that's coming :)

More will be announced publicly in the coming weeks as we finish the first pass at our docs and comms. But we're working out in the open - intrepid souls can look through the [project board](https://github.com/orgs/grafana/projects/108), or work their way down from [this issue](https://github.com/grafana/grafana/issues/44450) to get a sense of where we're going.

---

"As-code" is the name we use for the eclectic grouping of work related to making the experience of working with Grafana from code as smooth, native, and reliable as possible. This certainly includes fully supporting GitOps-style workflows, but we think the other direction - propagating changes made in the Grafana UI back to disk - to be equally essential for our use case.

Efforts here are still in the foundations-laying phase, and we haven't done much to clearly advertise what we're doing to the public. But we're at a point where what we have fundamentally is usable, at least for certain limited goals. So, as part of improving our communication story, we think publishing a roadmap of our plans is important.

### Grafana 8.2
- Importable dashboard scuemata, with plugins composed into place for use. Having this available is the first key step towards supplanting grafonnet with CUE
- Validation of dashboards by the dashboard schema usually works. This is checked as part of our CI process by [validating](https://github.com/grafana/grafana/blob/df81676d9ea9890bb9023ba5cc9f918694042165/pkg/schema/load/load_test.go#L56) the [dashboards in devenv](https://github.com/grafana/grafana/tree/main/devenv/dev-dashboards)
	- A guide for Grafana developers/contributors to updating devenv dashboards, given this new validation
- Rudimentary automatic generation of Typescript types through the [cuetsy](https://github.com/grafana/cuetsy) library. (Shipped as `grafana-cli cue gen-ts`, though results of codegen are still rough and have not yet replaced hand-written `*.gen.ts` files across the tree)

### Grafana 8.3

- Trimming and apply defaults: given some dashboard JSON (or any schematized object type, as we add more), remove all fields from the JSON that are schema-specified default value, as they can be considered implicit.
  - Been available behind a feature flag since 8.0, but had a key bug in the way that trimming worked over disjunctions in schema that just deletes all panels in a dashboard: #39412
- Just like there's a general scuemata family shape for [panel plugins](https://github.com/grafana/grafana/blob/main/cue/scuemata/panel-plugin.cue) that governs how a panel plugin is expected to define its schema in its `models.cue` file (ex: [timeseries panel](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/timeseries/models.cue)), we'll have a definitions for families to be used by datasource plugins.
  - We'll also likely have the scuemata written for some of the main datasource plugins - e.g., a `public/app/plugins/datasource/prometheus/models.cue`.
  - POC of the overall shape #37808
- Datasource type field now exists in dashboard JSON, and schema: #33817. (This is prerequisite to being able to schematize them.)
- Scuemata, which is fully generic and independent of Grafana, will be separated into its own repository.
  - Before that happens, we have to settle naming: #39013
- One of the most important things we want to enable with this schematization effort is GitOps-style flows. By 8.3, we're hoping to have an official prototype together.
- Public-facing documentation for using both scuemata and the schemas we provide
- All [hacky #-defs](https://github.com/grafana/grafana/blob/4c8c2f6c960094c8a5d36417d91b133e30e6275d/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L83-L125) in dashboard scuemata removed that were introduced in the [mad dash](https://github.com/grafana/grafana/pull/37857) to achieve basic validation.
- Object headers, scuemata version numberings in artifacts clarified. (This is kinda deep in the weeds to explain, so i'm leaving it as-is)

### Useful references

While we're working on fleshing out documentation for the 8.3 release, here are some useful pointers for _very_ industrious folks to explore:

- Old and dense, but still mostly accurate [overview of scuemata](https://docs.google.com/document/d/1qp3flLb6q1dPqTmHQ6MV-o6lOmlctfMNpzTx2MdzucY/edit#heading=h.ps08c938afna), the system we're using to write schema
- The [canonical base dashboard schema](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue) is here. "Base" means without any plugin-defined schema composed in. The "dist" version - which includes all plugins composed in - is [here](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dist/family.cue). We're not promising its filesystem location/import URL yet, but will make a promise around that by 8.3.
- If you want to follow progress, most as-code-related issue end up being tagged with [area/scuemata](https://github.com/grafana/grafana/issues?q=is%3Aopen+is%3Aissue+label%3Aarea%2Fscuemata).

**Comment by achew22** [comment]
Is there any chance the `as-code` effort is going to help build bindings for other languages beyond CUE? For example TypeScript, GoLang, or Python?

**Reply by sdboyer** [reply]
Excellent question! Given our general lack of docs until 8.3, i'll answer in more detail than is strictly necessary, as this question is getting a fair bit of eyeballs.

First, i have to flip part of your question around:

> bindings for other languages beyond CUE

The schema written in CUE aren't bindings. They _are_ - or at least, will be, once the system is fully in place, currently aiming at Grafana 9 - the canonical schema. All the other languages that might want to interact with the CUE schema are the bindings. 

This is one of the larger practical challenges with this effort, as it means a fundamental change to the way that the Grafana frontend application itself gets written. Historically, the dashboard model has really just been an incidental side effect of the way that the frontend _happened_ to be written - we like to say, dashboard JSON is a private API that ended up becoming very, very public. Consequently, every dashboards-as-code effort - including those maintained by Grafana Labs, like grafonnet - have been hacky reverse engineering efforts. Thus, the fact that these new schema are canonical, _not_ reverse engineerings, is one of the most important properties of the system, despite being entirely invisible.

The intended net effect of this shift in canonicality, combined with the guarantees around backwards compatibility/migratability afforded by scuemata, is that you have a firm, reliable foundation on which you can do your as-code work, with the reasonable expectation that things will not silently break when you upgrade Grafana versions, or plugin versions.

> For example TypeScript, GoLang

Because we're taking this "CUE is the one source of truth" approach, and because Grafana is written in TypeScript and Go, it's basically prerequisite to the success of the project that bindings for these languages exist. 

Crucial caveat - because dashboards are pluggable, the schema must necessarily be pluggable as well, which means that each language has to deal with making the schema for the plugins available in addition to the dashboard schema, and how they compose. ([The Matrix of Contents in the scuemata doc](https://docs.google.com/document/d/1qp3flLb6q1dPqTmHQ6MV-o6lOmlctfMNpzTx2MdzucY/edit#heading=h.7mt72zf53p82) also has a lot of this info.) In CUE, we do that composition for you in [this file](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dist/family.cue).

**Go**

Go is the main place that we currently interact with the CUE schemas. The main interface for getting the schema is [`pkg/schema/load`](https://pkg.go.dev/github.com/grafana/grafana/pkg/schema/load), which is absolutely designed for use by external programs, **though 8.3 is the earliest possible release at which we will begin making backwards compatibility guarantees.** The base dashboard and dist dashboard schemas are available via loader functions, which load the CUE files that are [embedded](https://github.com/grafana/grafana/blob/main/embed.go) into the binary.

Once loaded, actual interaction with the schemas is done through the interface defined in [`pkg/schema`](https://pkg.go.dev/github.com/grafana/grafana/pkg/schema). That's the essence of generic scuemata, and will _definitely_ be moving out into its own repo, very soon.

If you're looking for Go structs that represent dashboards in order to enable writing pure Go programs that emit Grafana dashboards, like what [grabana](https://github.com/K-Phoen/grabana) enables, we don't have those yet. We do recognize the value of doing this for at least certain use cases, but are likely going to wait for CUE upstream's [gocode encoder](https://pkg.go.dev/cuelang.org/go@v0.4.0/encoding/gocode).

**TypeScript**

In the Go case, we really aim at working with the whole of scuemata, but it's kinda at arm's length - existing Go code is really about shuttling a `[]byte` to a `schema.CueSchema`. TypeScript is the opposite - our primary use case is making the frontend able to work on just the "current" version of schemas. Thus, the "bindings" target only the latest schema in a scuemata, and aren't stitched together into a convenient top-level interface for churning out dashboards "as code."

We use the [cuetsy library](https://github.com/grafana/cuetsy) for doing the CUE->Typescript translation. We just [merged](https://github.com/grafana/grafana/pull/39694) the very first pass at the tool earlier this week - it's still rough and bug-laden! 

**Python**

Our plans are all towards satisfying the needs that we have to continue to create Grafana itself, and exposing those contracts in such a way that folks building externally can derive the same benefits. Such Dogfooding is really the only way to guarantee that this whole, enormous system stays precise and reliable.

Given that, and the fact that Grafana is not written in Python, we have no plans to directly support Python. If anyone's interested in doing that, though, i'd be happy to help them figure out how to best make our CUE ecosystem useful in Python :)

**Comment by JamesGibo** [comment]
Thanks for sharing the roadmap, I have been keeping an eye out for any updates about this since it was announced as it will be great improvement on the current system!

I help to maintain a library called [grafanalib](https://github.com/weaveworks/grafanalib) which allows you to generate Grafana dashboards using Python, so I am really keen to see how we can integrate this functionality into the library.

It would be great to chat about how this work is going in more detail, how we might be able to integrate this functionality into grafanalib and help you develop the schema. Please get in contact if you would to know more.

**Reply by sdboyer** [reply]
I'd love to talk! DM me on [twitter](https://twitter.com/sdboyer), we can find a time to chat it over

**Comment by justinTM** [comment]
hey @sdboyer ! we are considering refactoring our CI/CD Grafana automation using the latest CUE/polly work, and I was wondering if we could talk ?

**Reply by sdboyer** [reply]
absolutely! send me a dm on twitter, or i also DMed you on the k8s slack

**Comment by Asgoret** [comment]
Where I can see CUE examples of how to create dashboards? I'm starting learning grafonnet and then find a thread that moving from jsonnet to cue and can't find any link via search

**Reply by sdboyer** [reply]
I need to update this roadmap, and ideally make it more of an actual roadmap populated by issues, and less hopium about "things we're trying to ship by a particular version."

unfortunately, we don't have examples of that written up. It's possible to do today from the CUE schema we've put out, but you'd have to stitch together a lot of your own tools to really make it happen.

The best I can point you to for examples is [Polly](https://GitHub.com/pollypkg/polly), but the examples in there are half-baked, and I'm sure at least a few won't pass validation. (our schemas don't work with `schemaVersion < 30`)

It remains the goal that we obviate the need for grafonnet, but it has been a long and windy path to make something *enough* better that there's a realistic chance we don't end up at the same place, needing to start again from ~scratch.

On that front, there is at least good news: I've put up https://GitHub.com/grafana/thema, which is a fully genericized version of the (CUE) framework we're using to express our dashboard schemas, (and maybe all schemas, as we start moving Grafana as a whole towards schematization). 

What's there won't help you with your question at all. Nonetheless, I term it "good news" because I think a lot of what's blocked us over the past nine months has been trying to simultaneously do something _well_, but also do it fast and forego this like clear docs and boundaries. For something that really requires the participation and understanding of a large group of people with diverse experience, that has not worked well. Fair to say that we're buckling down to what it will actually take to go far, together, now.

**Reply by gurudatta-carbon** [reply]
are there are any CUE examples for creating dashboards now ?

**Reply by jay7x** [reply]
I'd extend the previous question to "What is Grafana-team-recommended way of creating dashboards-as-a-code these days"? Everything looks broken/deprecated/half-baked from my point of view :( Not sure where to go if I need to implement DaaS right now.

**Reply by dylan-tock** [reply]
I will amplify @jay7x's expansion.

From the outside looking in, there has been no official, supported way to generate Dashboards as Code (aka DaC) for an entire version of Grafana.  From the [breadcrumbs](https://github.com/grafana/grafana/issues/44450) [of](https://github.com/grafana/thema/issues/8) [issues](https://github.com/grafana/grafana/issues/44233) [and](https://github.com/grafana/grafana/issues/44160) [discussions](https://github.com/grafana/grafonnet-lib/issues/332) (but woefully few actual announcements) that you can find if you go far enough down the rabbit hole, it seems like the ability to create DaC for the eponymous and most popular product of the company is blocked behind needing a tool that can [track the ancestry of schema objects over multiple revisions](https://github.com/grafana/thema) that is [discovering](https://github.com/grafana/thema/issues/8) [its way](https://github.com/grafana/grafana/pull/49805) [like](https://github.com/grafana/grafana/pull/49373) [a](https://github.com/grafana/grafana/pull/44433) [colonial](https://github.com/grafana/grafana/issues/55769) [explorer](https://github.com/grafana/grafana/issues/56492). . . something that honestly sounds like a pet research project from a PhD/Principal Engineer.  The reason for needing such an ancestry tracker has never really been clear other than "there are back-end things that it will help with".  Meanwhile, [pre-existing methods of DaC](https://grafana.com/blog/2020/02/26/how-to-configure-grafana-as-code/) [are deprecated](https://github.com/grafana/grafonnet-lib/issues/332) none [of](https://github.com/grafana/dashboard-spec/tree/_gen) [the work](https://github.com/grafana/grafonnet-lib) for [those methods](https://github.com/grafana/grafonnet-lib/tree/master/grafonnet-7.0) is [usable](https://github.com/grafana/dashboard-spec/tree/master#code-generator) by subsequent grafana versions.

@torkelo took time in his portion of the GrafanaCon 2022 keynote to spotlight the "Go Git It" hackathon project for storing and reading files directly from git (and said Git Integration is "next for v9") and that @nopzor1200 explicitly called out the Git Integration as "exciting", but I did not see any other mention of automating the creation/maintenance of dashboards. I'm hope there will soon be improvements in the DaC area, but the continued silence (from @sdboyer and the rest of the Grafana team) does not make me enthusiastic. . . .

So.... yeah, some news on where you're going, what the milestones are along the way, and if we'll need to wait for Grafana 10 before we can have a "Grafana-team-recommended way of creating dashboards-as-code" would be a good thing™️ and most welcome.

**Reply by frossano-grafana** [reply]
Hi @dylan-tock 
I am Fabrizia, the new PM for the whole As-code functionality in Grafana, and I joined about a month ago. 
First, I want to reassure you that your concerns and criticism are shared internally, and that we are taking steps to address them.
Here at Grafana we really care bout having a cohesive ecosystems out for Dashboards as Code, to make sure the project is ready not just from a technical standpoint, but also from a usability one. You correctly pointed to the rabbit hole you had to dig through to find some information: well, this is exactly what we want to avoid, and it is taking us a bit longer than expected. 
As companies grow, one of the side effects is that it might take longer to align on a common direction, and I would like to make clear that Dashboard as Code is a priority for Grafana in the next months. Simply the fact that I am here and looking at this as a product with a dedicated team should be a testament to that. :) 
And yes, we’ll publish milestones and recommendations, just bear with us a little longer!

**Reply by jacksongoode** [reply]
Just to note on this thread that the [recent post made by the Grafana team on DaC](https://grafana.com/blog/2022/12/06/a-complete-guide-to-managing-grafana-as-code-tools-tips-and-tricks/) doesn't mention CUE or Grafonnet (I would point to [this fork as an updated version](https://github.com/rhowe/grafonnet-lib/) supporting v9) as an officially supported solution. If there was any time to provide some forward looking plan, it would be great to append some of the above discussion to a post like that. Excited for the future for DaC but generally sad that this hasn't been a priority.

**Reply by dylan-tock** [reply]
> Just to note on this thread that the [recent post made by the Grafana team on DaC](https://grafana.com/blog/2022/12/06/a-complete-guide-to-managing-grafana-as-code-tools-tips-and-tricks/) doesn't mention CUE or Grafonnet (I would point to [this fork as an updated version](https://github.com/rhowe/grafonnet-lib/) supporting v9) as an officially supported solution. If there was any time to provide some forward looking plan, it would be great to append some of the above discussion to a post like that. Excited for the future for DaC but generally sad that this hasn't been a priority.

I read the blog post and looked at all of the tools that @ishanjainn recommended. Of the five tools mentioned, only one (Grizzly) does dashboard creation[*]. And none of the examples in the [Grizzly repository](https://github.com/grafana/grizzly/tree/master/examples) have any real complexity to them.  The [dashboard example](https://github.com/grafana/grizzly/blob/master/examples/dashboard-simple.libsonnet) creates a dashboard with no rows or panels. Is creating the Dashboard step one and [step two](https://knowyourmeme.com/memes/how-to-draw-an-owl) is "Add the panels to your Dashboard" and they skipped that bit because it's just [adding small details](https://knowyourmeme.com/photos/572093-how-to-draw-an-owl)?

I'm not knocking Grizzly, but following links from its repo, you'll quickly end up here[**].  And I would have expected @ishanjainn to know that since they've done a [fair](https://github.com/grafana/thema/pull/75) [bit](https://github.com/grafana/thema/pull/78) [of](https://github.com/grafana/thema/pull/87) [work](https://github.com/grafana/thema/pull/88) with thema recently and, one would think, knew it was the presumed successor to Grafonnet/jsonnet.

@frossano-grafana, I hope you are having an enjoyable holiday. I presume you are on holiday since, given your position as "PM for the whole As-code functionality in Grafana", a blog post titled "A complete guide to managing Grafana as code: tools, tips, and tricks" would have been brought to your attention prior to publication and, were it to point to outdated resources that were announced as deprecated over a year ago, it would have been held back for a re-write to include new information regarding the steps forward.

I will also say, I'm glad Grafana has a PM that is looking at Dashboard as Code. I would be happier to see tangible results from that PM's work such as roadmaps, non-trivial implementation examples, and status updates.  I retain hope for this to come in the New Year.

[*] Technically, the Terraform provider *does* let you manage dashboards, but only by uploading dashboard JSON, not by using any of the terraform features.
[**] For dashboard configuration using Grizzly, the post explicitly links to [documentation](https://grafana.github.io/grafonnet-lib/usage/#grizzly) on the [Grafonnet-lib](https://grafana.github.io/grafonnet-lib/) website which [links](https://github.com/grafana/grafonnet-lib/tree/master/grafonnet-7.0) to the [grafonnet-lib](https://github.com/grafana/grafonnet-lib) repo which has [Issue 323](https://github.com/grafana/grafonnet-lib/issues/332), the bottom of which brings you here.

**Reply by JamesGibo** [reply]
> I read the blog post and looked at all of the tools that @ishanjainn recommended. Of the five tools mentioned, only one (Grizzly) does dashboard creation[*]. And none of the examples in the [Grizzly repository](https://github.com/grafana/grizzly/tree/master/examples) have any real complexity to them. The [dashboard example](https://github.com/grafana/grizzly/blob/master/examples/dashboard-simple.libsonnet) creates a dashboard with no rows or panels. Is creating the Dashboard step one and [step two](https://knowyourmeme.com/memes/how-to-draw-an-owl) is "Add the panels to your Dashboard" and they skipped that bit because it's just [adding small details](https://knowyourmeme.com/photos/572093-how-to-draw-an-owl)?
 
@dylan-tock If you are looking for Dashbaords as Code then please take a look at a library called [grafanalib](https://github.com/weaveworks/grafanalib) which allows you to generate Grafana dashboards using Python.

**Reply by dylan-tock** [reply]
> @dylan-tock If you are looking for Dashbaords as Code then please take a look at a library called [grafanalib](https://github.com/weaveworks/grafanalib) which allows you to generate Grafana dashboards using Python.

I am looking for Dashboards as Code that is officially supported by Grafana and will work with new versions of Grafana as they're released.  Or, barring that, a roadmap to something like that from @frossano-grafana when they're able to provide it.  Third-party libs in their "[very early stages](https://github.com/weaveworks/grafanalib#support)" aren't really a substitute for that.  I would imagine I'm not the only one following this conversation who feels similarly.

**Comment by varac** [comment]
@frossano-grafana  I'd love to hear some Grafana's official update on this topic. I appreciate the [blog post how to deploy resources as code](https://grafana.com/blog/2022/12/06/a-complete-guide-to-managing-grafana-as-code-tools-tips-and-tricks/), but that lacks a recommendation how to template i.e. dashboards.
In this article, Grizzly is featured with 

> Grizzly can also deploy dashboards built in Jsonnet using Grafonnet. (Learn more in the [Grafonnet documentation](https://grafana.github.io/grafonnet-lib/usage/#grizzly).)

When I visit the [grafonnet-lib docs](https://grafana.github.io/grafonnet-lib/), it states (already for a looooong time): 

> We're in the process of introducing generated code that can be used instead of the manually maintained Jsonnet code in the grafonnet directory. The generated code lives in grafonnet-7.0. It's generated from a new project, [grafana/dashboard-spec](https://github.com/grafana/dashboard-spec). The generated code is still incomplete, however, the components present are useable. We very much appreciate contributions in [grafana/dashboard-spec](https://github.com/grafana/dashboard-spec) for components yet to be added.

When I visit the linked [grafana/dashboard-spec repo](https://github.com/grafana/dashboard-spec) I see a deprecation warning: 

> DEPRECATED: Note, this repo is no-longer under active development. Whilst it was used to render Grafonnet's library for Grafana 7.0, future schema work is happening inside Grafana itself.

This is quite a mess for a while already.
Please give us an update, with a recommended approach for dashboards-as-code, which doesn't link to deprecated resources. Thanks !

**Comment by dylan-tock** [comment]
# Disclaimer/Background
A couple of notes before I get into the reply. My ultimate desires are: 1) The ability to create Dashboards with Config-As-Code via methods which are officially supported by Grafana and that will be updated as new versions of Grafana are released[^1]; and 2) Better communication from the Grafana Team regarding progress on the first desire.

I understand quoting individual members of the Grafana organization may have the effect of incentevizing them to reduce communication. However, it's the lack of communication that makes searching through Slack threads to get information necessary and having periodic updates regarding the status of "As Code" efforts would make these individual quotes either uninteresting (because it repeated previously available information) or unnecessary (since that information would be publicly available and linked to in an appropriate FAQ).  I explicitly want to say, the participation of @malcolmholmes and @duologic on the Grafana Community Slack is valuable and I would urge them to continue doing so.

[^1]: With the caveat that the tooling is more than "Manually create your Dashboard JSON, then use tools to deploy that bespoke JSON".

# Information
With the above context, here are some bits of info I found after seeing the previous comment and realizing it had been four months since @frossano-grafana posted in this Discussion thread.  The Grafana Community Slack server only stores threads up to 90 days. Because of this, I'm quoting the posts I'm linking to and providing context when necessary.

@malcolmholmes [on 3 Feb in response](https://grafana.slack.com/archives/C024B5R8T60/p1675458900537209?thread_ts=1675458810.047459&cid=C024B5R8T60) to the question "could someone point me to a resource to get started with developing dashboards with code? I have a library of panels that I would like to merge into a single dashboard reliably. (choosing different panels for custom dashboards)":
> "There are a range of ways. But Grafonnet is the most used as far as I know, and the one wr are likely to update soonest."

@malcolmholmes [5 minutes later in the same Slack thread](https://grafana.slack.com/archives/C024B5R8T60/p1675459216285979?thread_ts=1675458810.047459&cid=C024B5R8T60):
> "Just bear in mind, we will soon release an incompatible schematised version soon."

@malcolmholmes [on 28 Feb](https://grafana.slack.com/archives/C024B5R8T60/p1677618424769309?thread_ts=1677618258.313929&cid=C024B5R8T60)
> "Depends what you mean. If you just want to upload dashboards, Terraform will work for you. If you want to do limited composition of dashboards, it may still work. If you want to get more sophisticated, you will want something richer. Best at the moment is Jsonnet/Grafonnet. To deploy jsonnet you can use Grizzly. There is also a Jsonnet provider for Terraform that I haven't tried that could tender jsonnet which terraform deploys."

@malcolmholmes [on 3 Mar 2023 when asked for an ETA on when grafonnet updates would be public](https://grafana.slack.com/archives/C0146SMJ1FE/p1677850434162619?thread_ts=1676304377.280359&cid=C0146SMJ1FE):
> "We're working that out. But I don't expect it to be that long - in the low months.
> 
> What I have been advising people: learn jsonnet, it will still be relevant. Write systems using existing Grafonnet - the new one will still be rendering the same JSON, so won't be totally different. But avoid creating huge amounts of code, as the code will need some refactoring to bring it over to the new approach."

# My Conclusions
Based on the above quotes I've come to a few conclusions, a few that are encouraging and one which is disapointing. I'll list them here

Encouraging:
* The Grafana organization has been working on dashboard config as code.
* The config as code will be more than the "Write bespoke JSON/Copy dashboard JSON from the UI, then use terraform to deploy it".
* If you've used Grafonnet before, that experience will not go to waste.

Disappointing:
* The Grafana organization does not want to share this information publicly. I based this on:
    * no updates in this discussion by the "As-code PM"
    * while there is explicit acknowledgement about the current state of affairs (notably in the "Why is this important?" section of [this PR](https://github.com/grafana/grafonnet-lib/pull/376)), the only community engagement on the issue has been on a 1-on-1 basis in Slack/in person
    * movement of [PRs](https://github.com/grafana/grafonnet-lib/pull/380#issuecomment-1422310405) from the public grafonnet-lib repo to a private repo ([possibly the `grok` repo]](https://github.com/grafana/grok)) ostensibly to "reduce confusion" indicating that "reducing confusion" by hiding things is fine.
    * Acknowledgement that there is incorrect information currently published, but that updating those can happen ["once we are ready with our new version"](https://grafana.slack.com/archives/C024B5R8T60/p1676289756965489?thread_ts=1675458810.047459&cid=C024B5R8T60) indicating that "reducing confusion" by updating outdated and known incorrect information is not a priority.

I hope I'm wrong about the Grafana organization and what it is doing to incentivize the seemingly secretive behaviors of some of its members. I would really like to be proven wrong on this. However, I'm still waiting for those "milestones and recommendations" promised several months ago.

**Reply by Duologic** [reply]
Hey @dylan-tock (and other above)

Thanks for reaching out. I'm sorry you feel we're lacking in communication.

I just want to make clear that we are not trying to be secretive, on the contrary, we try to share early to help set expectations. However it is not easy to get that right, there are many people involved (many developers, community, customers, ...) and releasing something in public (even in an alpha state) comes at a cost (providing support, serving feedback loops, bugs and feature requests). Considering the scale we are aiming at, that cost can get amplified.

We're really close to getting something out there, hold on for a little more, please.

**Reply by AsocPro** [reply]
@Duologic I don't want to come across as impatient as I'm cool with waiting as I trust the grafana project to release when things are ready to go.
That being said I do want jump in with this feature once it is released. I have been watching the Grafana Labs blog as well as coming back here and am just wondering is this something that will come with a big hard to miss announcement or do I need to be watching for it in somewhere specific?

**Reply by Duologic** [reply]
Malcolm's message below is the latest we have on this.

Anything specific you are looking for?

**Reply by AsocPro** [reply]
I might be just confusing myself/misunderstanding a little bit. The main clarification I am looking for is: is the updated Grafonnet mentioned below now considered a future safe approach to creating dashboards? Previously it had been mentioned that large amounts of code should be avoided. I had been holding off on migrating a few larger projects until this was no longer the case.

**Reply by malcolmholmes** [reply]
github.com/grafana/grafonnet is the new version, derived from schemas, and is intended for active use. Feedback very welcome on that repo.

**Comment by malcolmholmes** [comment]
It is worth noting that we have now released an updated Grafonnet (https://github.com/grafana/grafonnet) that will get updates as Grafana is updated. Coverage is not complete as there's a lot of components (panels/datasources/etc) out there. I don't expect this to be the last language supported.

**Comment by Duologic** [comment]
I'm closing this issue as we have moved on from these topics, we keep on expaning our tooling to support 'as code's configuration management. If you have questions about this, please refer to the usual support channels.
